systemd/systemd.go

```diff
@@ -232,90 +234,103 @@ func (c *Collector) collect(ch chan<- prometheus.Metric) error {
 		return errors.Wrap(err, "could not get list of systemd units from dbus")
 	}
 
-	c.logger.Debugf("systemd getAllUnits took %f", time.Since(begin).Seconds())
+	c.logger.Debugf("systemd ListUnits took %f", time.Since(begin).Seconds())
 	begin = time.Now()
 	units := filterUnits(allUnits, c.unitWhitelistPattern, c.unitBlacklistPattern)
 	c.logger.Debugf("systemd filterUnits took %f", time.Since(begin).Seconds())
 
-	for _, unit := range units {
-		logger := c.logger.With("unit", unit.Name)
+	for unit := range units {
```

Comment:
This won't work as nothing waits for goroutine completion.  i.e. you will launch X goroutines and return nil immediately

 I think your previous iterationg with waitgroups was correct :)

Reply:
D'oh! I totally forgot to make a goroutine inside `filterUnits` and call close. Oops!